pallene/translator.lua

```diff
@@ -85,16 +85,87 @@ function Translator:translate_decl(decl)
+        self:translate_exp(exp.exp)
+    elseif tag == "ast.Exp.Concat" then
+        for _, inner_exp in ipairs(exp.exps) do
+            self:translate_exp(inner_exp)
+        end
+    elseif tag == "ast.Exp.Binop" then
+        self:translate_exp(exp.lhs)
+        self:translate_exp(exp.rhs)
+    elseif tag == "ast.Exp.CallFunc" then
+        self:translate_exp(exp.exp)
+        for _, arg in ipairs(exp.args) do
+            self:translate_exp(arg)
+        end
+    elseif tag == "ast.Exp.Cast" then
+        self:translate_exp(exp.exp)
+        if exp.target_start_loc then
+            print('Came here')
+            self:add_whitespace(exp.target_start_loc.pos, exp.target_end_loc.pos - 1)
+        end
+    elseif tag == "ast.Exp.Paren" then
+        self:translate_exp(exp.exp)
+    end
+end
+
 function Translator:translate_stat(stat)
-    if stat._tag == "ast.Stat.Decl" then
-        for _, decl in ipairs(stat.decls) do
-            self:translate_decl(decl)
+    if stat._tag == "ast.Stat.Block" then
+        for _, s in ipairs(stat.stats) do
+            self:translate_stat(s)
+        end
+    elseif stat._tag == "ast.Stat.While" then
+        self:translate_exp(stat.condition)
+        self:translate_stat(stat.block)
+    elseif stat._tag == "ast.Stat.Repeat" then
+        self:translate_stat(stat.block)
+        self:translate_exp(stat.condition)
+    elseif stat._tag == "ast.Stat.If" then
+        local clause = stat
+        while clause do
+            if clause._tag == "ast.Stat.Block" then
+                self:translate_stat(clause)
+                clause = nil
+            else
+                self:translate_exp(clause.condition)
+                self:translate_stat(clause.then_)
+                clause = clause.else_
+            end
         end
     elseif stat._tag == "ast.Stat.For" then
         self:translate_decl(stat.decl)
-    elseif stat._tag == "ast.Stat.Block" then
-        for _, s in ipairs(stat.stats) do
-            self:translate_stat(s)
+        self:translate_exp(stat.start)
+        self:translate_exp(stat.limit)
+        if stat.step then
+            self:translate_exp(stat.step)
+        end
+        self:translate_stat(stat.block)
+    elseif stat._tag == "ast.Stat.Assign" then
+        for _, exp in ipairs(stat.exps) do
+            print(require('inspect')(exp))
+            print('--')
+            self:translate_exp(exp)
+        end
+    elseif stat._tag == "ast.Stat.Decl" then
+        for _, decl in ipairs(stat.decls) do
+            self:translate_decl(decl)
+        end
+        for _, exp in ipairs(stat.exps) do
+            self:translate_exp(exp)
+        end
+    elseif stat._tag == "ast.Stat.Call" then
+        self:translate_exp(stat.call_exp)
+    elseif stat._tag == "ast.Stat.Return" then
+        for _, exp in ipairs(stat.exps) do
+            self:translate_exp(exp)
+        end
+    else
+        if stat._tag ~= "ast.Stat.Break" then
+            error("impossible")
```

Comment:
We would normally write this as 

```
elseif stat._tag == "ast.Stat.Break"
    -- Nothing
else
    error("impossible")
end
```

Reply:
Oh. I actually wrote that way and changed it. :')